### ISSUE TITLE
ocamltest incorrectly reports skipped tests as passing

### ISSUE THREAD

**Issue by shindere** [ocaml/ocaml#13481]
ON latest trunk (48ae66fabdd3d28ad6b6f741832a33b9aeb562c5), on a Linux
systme where gdb is _not_ installed, run:

```
./configure && make -j
./ocamltest/ocamltest.opt testsuite/tests/native-debugger/linux-gdb-amd64-test.ml
```

The last command will produce the following output:

```
 ... testing 'linux-gdb-amd64-test.ml' => passed
```

this output is misleading since the test has actually not been run at all,
since gdb is not installed; The test should be reported as having been
skipped.

I suspect this behaviour is an unintended consequence of #12895, so I am
wondering whether you @nojb would be willing to investigate this? Otherwise
I can certainly give it a try.

Finally, I have been hinted to this problem by @dra27 so all the credits
should go to him. Also, I think this problem has bitten @tmcgilchrist while
he was working on #13199. My guess is that, because of the present issue, it
was not noticed that the CI was successful not because the code was working
as expected, but rather because the tests were actually not run on the
GitHub workers since they do likely not have gdb / lldb.

**Comment by shindere** [comment]
#13482 proposes a fix for this issue.

**Comment by dra27** [comment]
Fixed in #13482, thanks @shindere!

**Comment by shindere** [comment]
David Allsopp (2024/09/26 10:36 -0700):
> Fixed in #13482, thanks @shindere!

Oh, I completely forgot to close this, thanks for your attention,
@dra27!

**Comment by tmcgilchrist** [comment]
This change is not working how I would expect.

Running a single test (that is being executed because it fails when I change the reference file) yields a skipped test, when it should be passed:

```
$ make one TEST=tests/native-debugger/linux-gdb-arm64-test.ml 
 ... testing 'linux-gdb-arm64-test.ml' => skipped
make[1]: Entering directory '/home/tsmc/ocaml/testsuite'
make[1]: Leaving directory '/home/tsmc/ocaml/testsuite'
make[1]: Entering directory '/home/tsmc/ocaml/testsuite'

List of skipped tests:
    /linux-gdb-arm64-test.ml

Summary:
     0 tests passed
     1 tests skipped
     0 tests failed
     0 tests not started (parent test skipped or failed)
     0 unexpected errors
     1 tests considered
make[1]: Leaving directory '/home/tsmc/ocaml/testsuite'
```

Running everything in a directory also reports everything as skipped even though at least two of the tests are being executed:

```
$ make one DIR=tests/native-debugger
Running tests from 'tests/native-debugger' ...
 ... testing 'linux-gdb-amd64-test.ml' => skipped
 ... testing 'linux-gdb-arm64-test.ml' => skipped
 ... testing 'linux-gdb-riscv-test.ml' => skipped
 ... testing 'linux-lldb-amd64-test.ml' => skipped
 ... testing 'linux-lldb-arm64-test.ml' => skipped
 ... testing 'macos-lldb-amd64.ml' => skipped
 ... testing 'macos-lldb-arm64.ml' => skipped
make[1]: Entering directory '/home/tsmc/ocaml/testsuite'
make[1]: Leaving directory '/home/tsmc/ocaml/testsuite'
make[1]: Entering directory '/home/tsmc/ocaml/testsuite'


List of skipped tests:
    tests/native-debugger

Summary:
     0 tests passed
     7 tests skipped
     0 tests failed
     0 tests not started (parent test skipped or failed)
     0 unexpected errors
     7 tests considered
make[1]: Leaving directory '/home/tsmc/ocaml/testsuite'
```

**Comment by gasche** [comment]
Note: #13317 proposes an `ocamltest` action in ocamltest, to test the semantics of ocamltest scripts. Using it to add ocamltest tests in our testsuite would catch regressions such as these, and give us more confidence that their fix is correct.

**Comment by gasche** [comment]
I looked at this tonight. My conclusion is that the behavior implemented by #12895 is correct, or at least follows the current informal specification of ocamltest:
- print "failed" if at least one test action succeeded
- print "skipped" if all test actions were skipped
- print "passed" otherwise -- if at least one action passed, with possibly some actions skipped

For a test script that contains several predicates before the real "meat" of the test, this specification is counter-intuitive, because it will show "passed" as soon as some of the predicates passed. The linux-gdb-amd64-test starts with `native-compiler;`, so it will show "passed" for all non-failing runs that have the native compiler enabled.

There is no simple/obvious fix for the issue, for example it would not be an improvement to start showing "skipped" as soon as *some* action skipped, because many tests have two different branches different on the configuration (flambda or not?), so they always have skipped actions.

Test scripts have a tree structure, with a series of actions that must be performed in sequence first, and then a list of alternative choices whose success is independent. We could consider that a test really "passes" only when at least one "full run" through the test passed, a path from the root of the test to one of the leaves of the test tree.

**Comment by gasche** [comment]
I propose an implementation of this "at least one full path" criterion in #13488. Note that the summary will remain imperfect even with this approach, sometimes it will return "passed" despite having skipped some test branches, so you need to print the full log (`./ocamltest/ocamltest foo.ml -e`) to see everything that happened.

**Comment by shindere** [comment]
This issue should now hopefully be fixed by the merge of #13508, which
has been prefered over #13488.
